app/pages/Pdf/ServicePdfPage.tsx

```diff
@@ -0,0 +1,421 @@
+import React, { useState, useEffect, useMemo, useRef } from "react";
+import ReactMarkdown from "react-markdown";
+import { useParams } from "react-router-dom";
+import qs from "qs";
+import {
+  ServiceAttribution,
+  TableOfContactInfo,
+  TableOfOpeningTimes,
+} from "components/listing";
+
+import { Datatable, Loader } from "components/ui";
+import config from "../../config";
+import {
+  fetchService,
+  generateServiceDetails,
+  getServiceLocations,
+  Organization,
+  Service,
+} from "../../models";
+import { Address } from "../../models/Meta";
+
+import styles from "./styles.module.scss";
+
+export const ServicePdfPage = () => {
+  const { id } = useParams<{ id: string }>();
+  const [service, setService] = useState<Service | null>(null);
+  const [pdfSource, setPdfSource] = useState("");
+  const [mapImgSrc, setMapImgSrc] = useState<string | null>(null);
+  const details = useMemo(
+    () => (service ? generateServiceDetails(service) : []),
+    [service]
+  );
+
+  const ref = useRef<HTMLDivElement>(null);
+
+  useEffect(() => {
+    if (id) {
+      fetchService(id).then((s) => {
+        setService(s);
+      });
+    }
+  }, [id]);
+
+  useEffect(() => {
+    if (mapImgSrc !== null && !pdfSource) {
+      // Map image has been fetched. We're now ready to make our request to the PDF conversion endpoint
+      fetch(`/api/services/html_to_pdf`, {
+        body: JSON.stringify({ html: ref.current?.outerHTML }),
+        method: "POST",
+        mode: "cors",
+        headers: { "Content-Type": "application/json" },
+      })
+        .then((resp) => resp.blob())
+        .then((blob) => {
+          setPdfSource(window.URL.createObjectURL(blob));
+        });
+    }
+  }, [mapImgSrc, pdfSource]);
+
+  if (!service) {
+    return <Loader />;
+  }
+
+  const { resource, recurringSchedule } = service;
+  const locations = getServiceLocations(service, resource, recurringSchedule);
+
+  if (locations.length > 0 && mapImgSrc === null) {
+    const { address_1, city, state_province, postal_code } =
+      locations[0].address;
+
+    // Use address rather than Lat/Lng as center. In some cases, a location may not have
+    // lat/lng or it may not be updated after address changes. Address seems to
+    // be a safer value to access.
+    const center = `${address_1}, ${city}, ${state_province}, ${postal_code}`;
+    const baseMapUrl = "https://maps.googleapis.com/maps/api/staticmap";
+    const params = {
+      center,
+      zoom: "14",
+      size: "375x225",
+      markers: `color:0xff6821|${center}`,
+      maptype: "roadmap",
+      key: config.GOOGLE_API_KEY,
+    };
+
+    fetch(
+      `${baseMapUrl}?${qs.stringify(params, { encodeValuesOnly: true })}`
+    ).then((mapResp) => {
+      setMapImgSrc(mapResp.url);
+    }).catch(() => {
+      // If there is an error, set the mapImgSrc to non-null so that the HTML to PDF API request will still be made
```

Comment:
I initialize the `mapImgSrc` as `null` given that I don't want the `PdfCrowd` API call to run _until_ the map image request has been completed, whether the response is either an error or success (that's why I used null rather than an empty string). Once the request returns the mapImgSrc is either an empty string or a valid url, and the PDF crowd effect will then run because the `mapImgSrc` is not null. 

Does this make sense? The order of various steps in this code may appear a little tangled. I was afraid of potential race conditions between the map/the PDF (and the `ref.current.outerHTML` for that matter), so if you see other potential problems with this let me know.